routers/api/actions/artifacts.go

```diff
@@ -0,0 +1,285 @@
+	m.Get(artifactRouteBase+"/{artifactID}/download", r.downloadArtifact)
+
+	return m
+}
+
+// withContexter initializes a package context for a request.
+func withContexter(ctx gocontext.Context) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(resp http.ResponseWriter, req *http.Request) {
+			ctx := context.Context{
+				Resp: context.NewResponse(resp),
+				Data: map[string]interface{}{},
+			}
+			defer ctx.Close()
+
+			ctx.Req = context.WithContext(req, &ctx)
+
+			next.ServeHTTP(ctx.Resp, ctx.Req)
+		})
+	}
+}
+
+type artifactRoutes struct {
+	prefix string
+	fs     storage.ObjectStorage
+}
+
+func (ar artifactRoutes) openFile(fpath string, contentRange string) (storage.Object, bool, error) {
+	if contentRange != "" && !strings.HasPrefix(contentRange, "bytes 0-") {
+		f, err := ar.fs.Open(fpath)
+		if err != nil {
+			return nil, false, err
+		}
+		_, err = f.Seek(0, os.SEEK_END)
+		return f, true, err
+	}
+	f, err := ar.fs.Open(fpath)
+	return f, false, err
+}
+
+func (ar artifactRoutes) buildArtifactUrl(taskID int64, artifactID int64, suffix string) (string, error) {
+	uploadURL := strings.TrimSuffix(setting.AppURL, "/") + strings.TrimSuffix(ar.prefix, "/") +
+		strings.ReplaceAll(artifactRouteBase, "{taskID}", strconv.FormatInt(taskID, 10)) +
+		"/" + strconv.FormatInt(artifactID, 10) + "/" + suffix
+	u, err := url.Parse(uploadURL)
+	if err != nil {
+		return "", err
+	}
+	return u.String(), nil
+}
+
+// getUploadArtifactURL generates a URL for uploading an artifact
+func (ar artifactRoutes) getUploadArtifactURL(ctx *context.Context) {
+	// get task
+	taskID := ctx.ParamsInt64("taskID")
+	task, err := actions.GetTaskByID(ctx, taskID)
+	if err != nil {
+		log.Error("Error getting task: %v", err)
+		ctx.Error(http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	artifact, err := actions.CreateArtifact(ctx, task)
+	if err != nil {
+		log.Error("Error creating artifact: %v", err)
+		ctx.Error(http.StatusInternalServerError, err.Error())
+		return
+	}
+	url, err := ar.buildArtifactUrl(taskID, artifact.ID, "upload")
+	if err != nil {
+		log.Error("Error parsing upload URL: %v", err)
+		ctx.Error(http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	ctx.JSON(200, map[string]interface{}{
+		"fileContainerResourceUrl": url,
+	})
+}
+
+func (ar artifactRoutes) uploadArtifact(ctx *context.Context) {
+	artifactID := ctx.ParamsInt64("artifactID")
+
+	artifact, err := actions.GetArtifactByID(ctx, artifactID)
+	if err != nil {
+		log.Error("Error getting artifact: %v", err)
+		ctx.Error(http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	itemPath := ctx.Req.URL.Query().Get("itemPath")
+	taskID := ctx.Params("taskID")
+	artifactName := strings.Split(itemPath, "/")[0]
+
+	if ctx.Req.Header.Get("Content-Encoding") == "gzip" {
+		itemPath += ".gz"
+	}
+	filePath := fmt.Sprintf("%s/%d/%s", taskID, artifactID, itemPath)
```

Comment:
If `itemPath` contains `../../../` , will it bypass its parent directories?

Reply:
itemPath is formatted as `{artifact_name/artifact_path}` in upload_artifact@v3 action. It validates `artifact_name` with safe rules. So it can not create name as `../../../` with danger paths.

Reply:
Will a hacker use `upload_artifact@v3`? Could a hacker construct the parameter manually?

Reply:
I add name validation method from github upload_artifact action to prevent danger artifact name.
https://github.com/go-gitea/gitea/pull/22738/commits/fcc396acafeda067effeade598eec6d4ae188349